### PR TITLE
clone mg_ptr for magic

### PR DESCRIPTION
**Pull request by atoomic**
Handle mg_ptr to avoid to share it with other PVs.

### CODE REVIEW

**Comment by garu** [comment]
@HaraldJoerg Thank you for reporting!
@atoomic Thank you for fixing!

Clone 0.42 should hit your friendly CPAN mirror any moment now :)

File: Clone.xs [hunk 1]
```diff
@@ -290,31 +290,54 @@ sv_clone (SV * ref, HV* hseen, int depth)
...
+
+        { /* clone the mg_ptr pv */
+          char *mg_ptr = mg->mg_ptr; /* default */
+
+          if (mg->mg_len >= 0) { /* copy the pv */
+            if (mg_ptr) {
+              Newxz(mg_ptr, mg->mg_len+1, char); /* add +1 for the NULL at the end? */
+              Copy(mg->mg_ptr, mg_ptr, mg->mg_len, char);
+            }
+          } else if (mg->mg_len == HEf_SVKEY) {
```
Comment (atoomic) [inline Clone.xs]: I realized that we have no unit tests to cover this part...

File: Clone.xs [hunk 2]
```diff
@@ -290,31 +290,54 @@ sv_clone (SV * ref, HV* hseen, int depth)
...
+            /* let's share the SV for now */
+            SvREFCNT_inc((SV*)mg->mg_ptr);
+            /* maybe we also want to clone the SV... */
+            //if (mg_ptr) mg->mg_ptr = (char*) sv_clone((SV*)mg->mg_ptr, hseen, -1);
```
Comment (atoomic) [inline Clone.xs]: maybe this is what we should prefer doing, the current solution above just do one SvRefCnt++

**Comment by garu** [comment]
Hey @atoomic! Looks like we broke Geo::Address::Formatter 😱 

https://rt.cpan.org/Ticket/Display.html?id=130195

Could you maybe take a look and see if we can fix it? If you have the time, of course!

**Comment by atoomic** [comment]
interesting, I will have a look at this
thanks for the update

**Comment by atoomic** [comment]
This is going to be fixed via https://github.com/garu/Clone/pull/26
Not sure if I can do a release of it, @garu can you review and publish a
release?

thanks
nicolas

**Comment by HaraldJoerg** [comment]
While you're at it: I stumbled over another segfault problem when a structure happened to contain a DBI database handle.  On Linux, the demo attached dies with Clone 0.41 and Clone 0.42, while it seems to work on Windows.

Cloning database handles is, of course, veeery suspicious (and should not have happened in my code in the first place), but some sort of safeguarding in Clone would be nice.
[dbhclone.txt](https://github.com/garu/Clone/files/3444009/dbhclone.txt)

**Comment by atoomic** [comment]
Any chance you could report that as a GitHub ticket, with what would be
nice a simple script to reproduce the error?

Any chance this is related to https://github.com/garu/Clone/issues/17

thanks in advance
nicolas

**Comment by HaraldJoerg** [comment]
Sure, will report an issue.  The relation to #17 is probably just what's given there as a root cause: "Database handles can't be cloned".

**Comment by atoomic** [comment]
ok I realized I've permissions to publish Clone,
I've just uploaded an updated version of Clone 0.43 to CPAN,
it should be available in the next minutes.

Please confirm that this is fixing your issue.

sincerely
nicolas

**Pull request merged**